ARTEMIS-1156: FIX: Long Autoboxing occurring on Hot Path

michaelandrepearce opened the pull request:
Building on ARTEMIS-905 JCtools ConcurrentMap replacement  first proposed but currently parked by @franz1981, replace the collections with primitive key concurrent collections to avoid auto boxing.

The goal of this is to reduce/remove autoboxing on the hot path.
We are just adding jctools to the broker (should not be in client dependencies)
Like wise targeting specific use case with specific implementation rather than a blanket replace all. 

We have to create our own primitive long concurrent hash set, based on jctools as jctools currently don't have one.

michaelandrepearce commented:
See Jira with profiler output showing long object creation caused by auto boxing [https://issues.apache.org/jira/browse/ARTEMIS-1156](url)

franz1981 commented:
@michaelandrepearce Good work and great findings!
My 2 cents on it, the new map is great but:
1) it has a slightly different behaviour about atomic operations (eg [ConcurrentHashMap::compute javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-)) hence please checks if there are callers relying on such behaviours
2)  it doesn't shrink/grow  like the vanilla ConcurrentHashMap -> when it grows it is not incremental by little chuncks, but could put a lot of pressure on Humongous Allocations hurting the heap allocations of byte[] performed during message translations    
3) JournalCompactor in JournalImpl::compac is using a keySet from it (producing the unwanted Longs

So it is important to make sure that the common usage do not make them happen or at least , that the benefits are well balanced by these drawbacks
Anyway, the PR is great and I'm pretty happy about it :)

michaelandrepearce commented:
@franz1981 
On 1) compute isn't being used within this class, the methods used are ::get ::put ::contains ::containsKey, indeed this is why i state this is targeted, instead of your original blanket replace all, each case should be evaluated and implemented on merit.

On 2) agreed, in warm up stages we would see this, but once warm and message rate through the broker, it would stabilise its size, this is why we don't see hashmap.entry being created heavily in the profiler once warm.

On 3) good spot. lets address that.

michaelandrepearce commented:
@franz1981 just updated to add in the journal compactor.

I ran out of time today, to deploy and run this with a profiler on our servers to verify that this hasn't pushed the auto boxing further down or else where, i will do this tomorrow.

In particular can you look at the way i have handled addAll for the NonBlockingHashSetLong, is there a better way?

Also for the HashMap that had long keys in the journal which is not concurrent, I've used the LongHashMap from netty.

clebertsuconic commented:
I want to look into this before merge.  I'm out of my home office now. I will be back in a bit today. 


Also. I want to spend time on reference counts on messages this next weeks and use pooled buffers on messages as well.

clebertsuconic commented on artemis-journal/src/main/java/org/jctools/maps/NonBlockingHashSetLong.java:
```diff
@@ -0,0 +1,105 @@
+ * A simple wrapper around {@link NonBlockingHashMapLong} making it implement the
+ * {@link java.util.Set} interface.  All operations are Non-Blocking and multi-thread safe.
+ */
+public class NonBlockingHashSetLong extends AbstractSet<Long> implements Serializable {
```
Isn't this class available on any jctools package? 
Can't we just reuse the library instead of copy & paste?

clebertsuconic replied:
@michaelandrepearce would be ok to move this under utils, under an artemis package? just in case.. I think that's more clear, especially with osgi usage.

michaelandrepearce commented:
No. Unfortunately not, there is an int version and a hashset for objects (but that will cause autoboxing)

michaelandrepearce commented:
it is largely based on the NonBlockingHashSet that internally uses NonBlockingHashMap taking object keys, but replaced the internal map with their NonBlockingHashMapLong and added the primitive methods.

clebertsuconic commented:
@michaelandrepearce can you fix the checkstyle

clebertsuconic commented:
@michaelandrepearce Oh.. never mind.. I can ammend during checkout.

michaelandrepearce commented:
Yeah still got a few bits to clean up. Also I need to run it on server under load I ran out of time today. But review comments would be good

michaelandrepearce commented:
Well there's another fix I have locally I need to commit to fix the distribution as it's missing the dependency in the zip/tar go. But on the move atm

michaelandrepearce commented:
dep.xml updated and check style fixed.

have managed to get results back from testing, allocation of longs is greatly reduced (oddly theres one autobox in JournalTransaction, ill have to check that again, but now at least significant sector is just the byte[] from the unpooled buffer.

michaelandrepearce commented:
<img width="1001" alt="screen shot 2017-05-10 at 19 04 05" src="https://cloud.githubusercontent.com/assets/1387822/25913734/b9930f5e-35b3-11e7-84af-a4073e0643c3.png">
<img width="1001" alt="screen shot 2017-05-10 at 19 03 02" src="https://cloud.githubusercontent.com/assets/1387822/25913736/b999c646-35b3-11e7-928f-8a64c9e70e64.png">

michaelandrepearce commented:
from 127mb prior and now just 11mb. :), and i think we can remove that also :) might have a look what the Integer is upto, in the allocations , it looks like it auto boxing also.

clebertsuconic commented:
@michaelandrepearce are you using replication? FileWrapperJournal used on replication has a HashMap<Long>

michaelandrepearce commented:
ok so the int, is due to connection.getID return Object, for the two implementations netty it is int, and InVM its a String UUID. I would suggest the getID becomes an int on the interfaces, and where the InVM uses currently UUID to regenerate a unique id, as its within a VM we can use a Sequence to generate a unique int.

This is a bit more invasive, so i won't to that in this PR.

clebertsuconic commented:
@michaelandrepearce +1 on changing the connection.getID() to a generated sequence.

It's a bit invasive though as ProtocolClientConnectionManager and other places will use the ID to remove failed connections... etc...

https://issues.apache.org/jira/browse/ARTEMIS-1158

michaelandrepearce commented:
So I'm just doing a bit more of a high load, seems the hash set, as unlike the maps used in the JournalImpl which are fairly permanent, the sets are temporal they're killing the outside TLAB allocation every time compaction runs. As such its better to keep them as normal sets i think.

@franz1981 Im probably going to revert the bit i added for point 3) you raised.

As its having a negative effect. I really just want to hit the hot path, and compaction is periodic (though frequent isn't as hot)

franz1981 commented:
It makes sense for me!! I've to check the code for the addAll and about the connections..there is another JIRA to be issued there: about using as NettyConnectionI.getID() the hashCode of  the Netty channel instead of Channel::id(). But for different reasons: if 2 channels shares the same hashcode (with big scale connections it could happen) you risk to pick the wrong one in the map because they are sharing the same key!!!!

Changing subject! Very pleased by the results....I was expecting to be good, but is even better, well done!
@clebertsuconic  @michaelandrepearce  Now IMHO it makes sense to reason if the max growth size of a primitive ConcurrentMap after different resizing ops is not too big to be retained in periods with low load...wdyt?

clebertsuconic commented:
@michaelandrepearce this should fix some of the boxing you had with connectionID:

https://github.com/clebertsuconic/activemq-artemis/commit/ddb391d26bcbc37e35f22d6cd22d6fcd86a555f7

clebertsuconic commented:
@michaelandrepearce I already pushed my commit into master...

I will have to hold this PR until tomorrow. I had a few test failures... I need to make sure before I merge this. But I'm having a good impression when even running a simple producer / consumer from the CLI. (together with the whole set of changes)

michaelandrepearce commented:
If you could hold of merging. I am wanting to see by changing the collections to ones used in Apache bookkeeper if we get any better.

michaelandrepearce commented:
@clebertsucnic 
Just seen what you did, oh I like the idea what you did instead of going to primitive, simply go the other way an make it an object and avoid constant autoboxing. 

I wonder if the long id, if we just kept it as a Long instead of long would avoid all the autoboxing and means we wouldn't need special collections. 

Let me try this also.

franz1981 commented:
@michaelandrepearce Beware that some autoboxing will disappear due to escape analysis from the JIT, but only if a method got compiled: using profilers sometimes could show instances that in a real server with properly compiled methods will not exists. The same for primitive arrays due to TLAB parsability of the heap: https://shipilev.net/jvm-anatomy-park/5-tlabs-and-heap-parsability/

michaelandrepearce commented:
this will only occur where the primitive:boxed value is cached, the jvm will cache/inline upto only a certain amount. If you have a constantly increasing number like in this case an ID it cannot do this magic.

If like in this case it is an ID rather than something used for arithmetic then object reference can be as good as a primitive. It would agreed be terrible if you did arithmetics on it.

The issue with as the code sits currently is that it is autoboxing multiple times an primitive currently and the id is outside the caching ranges because it seems ever increasing. As such multiple Long objects are being created. And destroyed a while later. 

After the nudge the Clebert gave with his id solution.

My thoughts are:
If we can either just make the object once and pass it around by reference or we can use primitive will solve this.

First approach we have tried is primitive and is safe if we ever do arithmetic on it. Though the complexity is higher (aka special collections).

michaelandrepearce commented:
```
public class Boxing {
    
    
    public static void main(String... args){
        
        //ensure we max'd out the caching magic.
        for (long i = 0; i < 100000000; i++){
            Long v = Long.valueOf(i);
        }
        
        int times = 1000000000;
        
        long startreference = System.nanoTime();
        reference(Long.valueOf(1000000001), times);
        long endreference = System.nanoTime();

        long startprimitive = System.nanoTime();
        primitive(1000000001l, times);
        long endprimitive = System.nanoTime();

        long startboxing = System.nanoTime();
        boxing(1000000001, times);
        long endboxing = System.nanoTime();


        long startunboxing = System.nanoTime();
        unboxing(Long.valueOf(1000000001), times);
        long endunboxing = System.nanoTime();



        System.out.println("reference   :" + (endreference - startreference));
        System.out.println("primitive   :" + (endprimitive - startprimitive));
        System.out.println("boxing      :" + (endboxing - startboxing));
        System.out.println("unboxing    :" + (endboxing - startboxing));

    }
    
    
    
    public static Long reference(Long number, int times){

        Long value = null;
        for(int i = 0; i < times; i++)
            value = number;

        return value;
    }

    public static Long boxing(long number, int times){

        Long value = null;
        for(int i = 0; i < times; i++)
            value = number;

        return value;
    }


    public static long primitive(long number, int times){


        long value = -1;
        for(int i = 0; i < times; i++)
            value = number;
        
        return value;

    }

    public static long unboxing(Long number, int times){


        long value = -1;
        for(int i = 0; i < times; i++)
            value = number;

        return value;

    }
}
```


```
reference   :1470275
primitive   :934418
boxing      :4394771377
unboxing    :4394771377
```

franz1981 commented:
@michaelandrepearce Got it, there is another effective solution to this: build an interner, similar to a  StringInterner, but with a different hashed logic in order to avoid the JVM duplicating autoboxed instances each time. 
Few lines of code to explain it better:

```
//it will retain in LRU fashion the last 64 Long instances 
LongInterner longCache = new LongInterner(64);
long id;

//instead of:
put(id , connection);

//interned instance
Long cachedId = longCache.intern(id);
put(cachedId, connection);
```
wdyt?

The caching/escape analysis are different things AFAIK: 
- caching is performed by the library on long->Long, and put by the compiler when autoboxing (eg [Long::valueOf](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/lang/Long.java#837)) 
- [escape analysis](http://docs.oracle.com/javase/7/docs/technotes/guides/vm/performance-enhancements-7.html#escapeAnalysis): newly created Long instances no longer exists on the heap but being unpacked on the stack if used temporary, but need to call by hand new Long(long), bypassing the cached logic and do not retain it (seems not our case anyway)

eg http://www.rationaljava.com/2015/12/how-long-does-it-take-jvm-to-effect.html

franz1981 commented:
@michaelandrepearce on the Long.valueOf code I've posted there are the limit of the caching of the java library:  `if (l >= -128 && l <= 127) { // will cache`

michaelandrepearce commented:
So we have:

1. primitive collections
1. id's as objects (Long in this case)
1. flywheel/interning caching pattern

For 1:
 we have tried, looks good as long as we just do hot paths, can very quickly over complicate our code, or start bringing negative effects.

For 2:
I haven't had a chance to see how much work, but this maybe simpler/cleaner imho. Though does lend its self to people in future accidentally introducing the boxing issue again.

For 3:
Should work nicely, but only upto a predefined cache size. e.g. what occurs if cache size is N and i end up with N+1000 longs, we end up in same position as with Long.valueOf. 
Also some complexities introduced, how big should the cache be? how does someone tune it, i run a very large vm, but fred runs a small one, this works for an app that you deploy just in your own DC, but I'm weary of setting a magic cache size where it gets deployed in many differing env's.

For this reason, id personally favour looking into option 2, then keeping a low scoped option 1.

franz1981 commented:
@michaelandrepearce 
Re 1:
agree with sorrow: I've to admit that an all-out primitive solution  would be ideal in the long term (in GC terms) , but with too many effects in the short one and...is a breaking change.
Re 2: 
agree, me too, I've to check the code better!
Re 3:
depends: giving its fixed size, collisions will exists but we know how we generates the ids :) and this will help to choose the hashed function wisely (is not a general purpose interner/cache).
But I agree on the size of the cache (I tends to size interners of about 16 elements, max 32 in general, to be more cache effective): avoid magic numbers/tuning is better.
And it will add complexities too!

michaelandrepearce commented:
So I've just had dived through a few flows very quickly using tracing, there is one key area if we change from long to Long and that is in the protocol/buffer exchanges.

Case: ReplicationStartSyncMessage, ::encodeRest ::decodeRest

Changing this will introduce new object's, and we will need to fully retest to make sure we simply won't  just move the problem (which i think we might).

michaelandrepearce commented:
@franz1981 As noted above, i wanted to look at the collections that bookkeeper uses, the benefit here with these seems to be much reduced outside tlab allocation of long[] when it has to resize/rehash. Whilst also bringing the benefit of much reduced tlab allocating of Long due to autoboxing.

https://github.com/apache/bookkeeper/tree/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/util/collections

downside is these collections are not in a nice dependency we can just bring in, as such I've got to do a copy paste job (hopefully not a big deal)

I think I'm happier with these because of the much reduced outside tab allocation.

franz1981 commented:
Have you taken a look to http://fastutil.di.unimi.it ?

michaelandrepearce commented:
correct me if I'm wrong, but these are not concurrent.

michaelandrepearce commented:
FYI:
https://issues.apache.org/jira/browse/BOOKKEEPER-964

franz1981 commented:
You're right I've missed the concurrent part :P

michaelandrepearce commented:
retest this please.

michaelandrepearce commented:
@franz1981 @clebertsuconic 
So all updated to use the collections from bookkeeper. ran several load tests through with and without changes, defiantly reduced the Long object creation caused by auto boxing, and also addressed the issue i saw in using JCTools variant. :)

Unfortunately cleanest way is to copy in the classes/tests we need, i have tried my best to keep it clean, and also had to obviously make changes to the class's to meet Artemis checkstyle rules etc. So all should be done etc etc. 

But i am at a loss to whats occurring with the build server and why failing, due to sevntu.

clebertsuconic commented:
I believe you need to add dependencies to features.xml. 

I had to do it on my branch yesterday. 

 My branch called 1263  had it.

michaelandrepearce commented:
But I've removed deps on jctools entirely. Let me check your branch.

michaelandrepearce commented:
So i see your change :), but i don't think the build issue isn't that.

As noted i have removed the dependency on jctools entirely, as ended up using the bookkeeper collections.

I'm able to build locally, so i am stumped a bit now.

michaelandrepearce commented:
error is:

```
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Unresolveable build extension: Plugin org.apache.felix:maven-bundle-plugin:3.3.0 or one of its dependencies could not be resolved: Failure to find org.apache.felix:maven-bundle-plugin:jar:3.3.0 in http://sevntu-checkstyle.github.io/sevntu.checkstyle/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of sevntu-maven has elapsed or updates are forced @ 
```

clebertsuconic commented:
Ok. I'm stuck with the iPhone only now. :)

I can look / run tests in the afternoon.  I can amend anything if needed.

michaelandrepearce commented:
No worries at all, if i have any divine inspiration i will let you know, if not thanks in advance 👍

michaelandrepearce commented:
Seems build has got past that issue. Doing nothing at all. Weird. Anyhow now it's building again looks like I accidentally added some white space in a stress test class I'll sort that.

clebertsuconic commented:
@michaelandrepearce I'm not sure what version i checked.. but I needed to add this to build:

https://github.com/clebertsuconic/activemq-artemis/commit/a01e88ece5e1af48f1e9ceb074edb456e4e951a1

clebertsuconic commented:
to check locally, you have to do:

```
mvn -Pdev install
```

michaelandrepearce commented:
@clebertsuconic yeah, checkstyle came back on the build, earlier just wasn't even doing anything and just consistently complained 

```
ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Unresolveable build extension: Plugin org.apache.felix:maven-bundle-plugin:3.3.0 or one of its dependencies could not be resolved: Failure to find org.apache.felix:maven-bundle-plugin:jar:3.3.0 in http://sevntu-checkstyle.github.io/sevntu.checkstyle/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of sevntu-maven has elapsed or updates are forced @ 
```

And this magically stopped and is building again now.

michaelandrepearce commented:
all green now :)

michaelandrepearce commented:
@clebertsuconic if you and @franz1981 want to have a final go over, once happy if you need me to let me know ill rebase and squash for you so it can be merged.

clebertsuconic commented:
@michaelandrepearce Compactor is failing pretty bad at the moment...

ConcurrentLongHashSet needs a minimal size of numSections...

```
[Thread-0 (ActiveMQ-IO-server-org.apache.activemq.artemis.core.server.impl.ActiveMQServerImpl$4@192d3247)] 15:25:35,912 ERROR [org.apache.activemq.artemis.journal] AMQ144003: Error compacting: java.lang.IllegalArgumentException
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:108) [guava-19.0.jar:]
	at org.apache.activemq.artemis.utils.collections.ConcurrentLongHashSet.<init>(ConcurrentLongHashSet.java:63) [:]
	at org.apache.activemq.artemis.utils.collections.ConcurrentLongHashSet.<init>(ConcurrentLongHashSet.java:59) [:]
	at org.apache.activemq.artemis.utils.collections.ConcurrentLongHashMap.keysLongHashSet(ConcurrentLongHashMap.java:181) [:]
	at org.apache.activemq.artemis.core.journal.impl.JournalImpl.compact(JournalImpl.java:1490) [:]
	at org.apache.activemq.artemis.core.journal.impl.JournalImpl$12.run(JournalImpl.java:1413) [:]
	at org.apache.activemq.artemis.utils.OrderedExecutorFactory$OrderedExecutor$ExecutorTask.run(OrderedExecutorFactory.java:101) [:]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [rt.jar:1.8.0_73]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [rt.jar:1.8.0_73]
	at java.lang.Thread.run(Thread.java:745) [rt.jar:1.8.0_73]

```


This is because the ConcurrentLongHashMap was checking for a list of IDs on an empty set. Which was fine before (we would iterate over nothing and give up the loop before).


I'm not sure yet what to do here. To change the minimal... or what to do.. I need to get my head around these collections.. so I appreciate any help? :)

clebertsuconic commented:
I was looking at ReplicatedFailoverTest for this. All of the tests are throwing that exceptions.

clebertsuconic commented:
@michaelandrepearce sent a PR to your branch. if that's ok squash that commit as yours.

michaelandrepearce commented:
Thanks :), I'm out for the evening tonight. But will look into firsts thing in the morning.

Sent from my iPhone

> On 11 May 2017, at 21:49, clebertsuconic <notifications@github.com> wrote:
> 
> @michaelandrepearce sent a PR to your branch. if that's ok squash that commit as yours.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

clebertsuconic commented:
@michaelandrepearce I have created a new branch called ARTEMIS-1156 on my fork in case that's easier since I had already rebased against master.

clebertsuconic commented:
I have a few extra commits. Like since you created a collections package, I'm considering moving all collections there.

clebertsuconic commented:
I had one very good run on the testsuite at my box for this branch:

https://github.com/clebertsuconic/activemq-artemis/commits/ARTEMIS-1156


if you agree with my fix on the HashSet we can go with that after cleaning up the commit message.

I would keep the commit to move the collections to the new package separate of course.


will wait for you before we merge anything.

michaelandrepearce commented:
@clebertsuconic, have patched in the hashset fix, thanks for this. I also made the same change on the hashmap. I have squashed and rebased.

I agree the move of the other collections should be a separate commit. I have your commit for this as a patch, but i don't know how to apply it, with you as the committer.

michaelandrepearce commented:
@clebertsuconic  Don't worry figured it out :) managed to commit the moving collections to the new collections package as you.

michaelandrepearce commented:
fyi, doing initial tests looks like still all good to me on the auto boxing of long front.

franz1981 commented:
@michaelandrepearce  ASAP I'll take a look,but seems good now! :+1:

clebertsuconic commented:
@michaelandrepearce merged... did some tweaks to preserve history on TypedProperties.. all merged now...


Awesome work on this!

michaelandrepearce commented:
@clebertsuconic great news. And def no worries on doing tweeks for git history, your git fu far beats mine :) 

thanks to you and also @franz1981 for the feedback, review, comments and discussion, it really helps.

franz1981 commented:
@clebertsuconic @michaelandrepearce you're welcome and thanks to you: I'm very very happy of this fix guys :+1:  Awesome work!
